### PR TITLE
[pigeon] Make dart options const

### PR DESCRIPTION
**Pull request by bparrishMines**
Fixes https://github.com/flutter/flutter/issues/120374

`webview_flutter_android` and `webview_flutter_wkwebview` both use the `ConfigurePigeon` annotation in their pigeon files:

https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_android/pigeons/android_webview.dart#L7
https://github.com/flutter/plugins/blob/main/packages/webview_flutter/webview_flutter_wkwebview/pigeons/web_kit.dart#L7

This requires that `DartOptions` be a const since the `PigeonOptions` is const. 

This looks like this would be a breaking change though. The plugins mentioned above won't be able to update to the latest version because of this, so it could also be considered a bug? Let me know if there is a better way to handle this.

Note that `JavaOptions`, `ObjcOptions`, `CppOptions`, `KotlinOptions`, and `SwiftOptions` are already immutable.

EDIT: (tarrinneal)

This has also been updated to revert a change making Java Flutter Api classes final as this was changed unintentionally.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by tarrinneal** [comment]
Since this is a breaking change, we should probably bump the version up to 8.0.0

File: packages/pigeon/pigeons/configure_pigeon_dart_out.dart [hunk 1]
```diff
@@ -4,7 +4,11 @@
 
 import 'package:pigeon/pigeon.dart';
 
-@ConfigurePigeon(PigeonOptions(dartOut: 'stdout', javaOut: 'stdout'))
+@ConfigurePigeon(PigeonOptions(
+  dartOut: 'stdout',
+  javaOut: 'stdout',
+  dartOptions: DartOptions(),
```
Comment (tarrinneal) [inline packages/pigeon/pigeons/configure_pigeon_dart_out.dart:10]: Should this be `const DartOptions`?

File: packages/pigeon/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,8 @@
+## 9.0.0
+
+* **Breaking Change** Updates `DartOptions` to be immutable and adds const to the constructor.
+* **Breaking Change** Reverts `final` changes to Flutter Api classes.
```
Comment (stuartmorgan) [inline packages/pigeon/CHANGELOG.md:4]: I don't think this one is breaking? Is there code that wouldn't compile due to something no longer being final?

**Review by stuartmorgan — APPROVED** [review]
LGTM

File: packages/pigeon/CHANGELOG.md (hunk 2 above)
Comment (tarrinneal) [reply packages/pigeon/CHANGELOG.md:4]: I guess it's more of an un-breaking change

**Pull request merged**